File: server/storage.ts
import { 
  users, type User, type InsertUser,
  posts, type Post, type InsertPost,
  comments, type Comment, type InsertComment,
  aiChats, type AiChat, type InsertAiChat,
  aiChatMessages, type AiChatMessage, type InsertAiChatMessage,
  documents, type Document, type InsertDocument,
  votes, type Vote, type InsertVote
} from "@shared/schema";

export interface IStorage {
  // User operations
  getUser(id: number): Promise<User | undefined>;
  getUserByUsername(username: string): Promise<User | undefined>;
  getUserByEmail(email: string): Promise<User | undefined>;
  createUser(user: InsertUser): Promise<User>;
  
  // Post operations
  createPost(post: InsertPost): Promise<Post>;
  getPost(id: number): Promise<Post | undefined>;
  updatePost(id: number, data: Partial<Post>): Promise<Post | undefined>;
  getPostsByCategory(category: string, page: number, limit: number, sortBy: string): Promise<Post[]>;
  countPostsByCategory(category: string): Promise<number>;
  searchPosts(query: string): Promise<Post[]>;
  searchPostsByCategory(query: string, category: string): Promise<Post[]>;
  
  // Comment operations
  createComment(comment: InsertComment): Promise<Comment>;
  getComment(id: number): Promise<Comment | undefined>;
  getCommentsByPostId(postId: number): Promise<Comment[]>;
  updateCommentVoteCounts(id: number): Promise<void>;
  
  // AI Chat operations
  createAiChat(chat: InsertAiChat): Promise<AiChat>;
  getAiChat(id: number): Promise<AiChat | undefined>;
  getAiChatsByUserId(userId: number): Promise<AiChat[]>;
  
  // AI Chat Message operations
  createAiChatMessage(message: InsertAiChatMessage): Promise<AiChatMessage>;
  getAiChatMessages(chatId: number): Promise<AiChatMessage[]>;
  
  // Document operations
  createDocument(document: InsertDocument): Promise<Document>;
  getDocument(id: number): Promise<Document | undefined>;
  updateDocument(id: number, data: Partial<Document>): Promise<Document | undefined>;
  getDocumentsByCategory(category: string): Promise<Document[]>;
  getAllDocuments(): Promise<Document[]>;
  deleteDocument(id: number): Promise<void>;
  
  // Vote operations
  createVote(vote: InsertVote): Promise<Vote>;
  getVoteByUserAndPost(userId: number, postId: number): Promise<Vote | undefined>;
  getVoteByUserAndComment(userId: number, commentId: number): Promise<Vote | undefined>;
  updateVote(id: number, data: Partial<Vote>): Promise<Vote | undefined>;
  deleteVote(id: number): Promise<void>;
}

export class MemStorage implements IStorage {
  private users: Map<number, User>;
  private posts: Map<number, Post>;
  private comments: Map<number, Comment>;
  private aiChats: Map<number, AiChat>;
  private aiChatMessages: Map<number, AiChatMessage>;
  private documents: Map<number, Document>;
  private votes: Map<number, Vote>;
  
  private userIdCounter: number;
  private postIdCounter: number;
  private commentIdCounter: number;
  private aiChatIdCounter: number;
  private aiChatMessageIdCounter: number;
  private documentIdCounter: number;
  private voteIdCounter: number;

  constructor() {
    this.users = new Map();
    this.posts = new Map();
    this.comments = new Map();
    this.aiChats = new Map();
    this.aiChatMessages = new Map();
    this.documents = new Map();
    this.votes = new Map();
    
    this.userIdCounter = 1;
    this.postIdCounter = 1;
    this.commentIdCounter = 1;
    this.aiChatIdCounter = 1;
    this.aiChatMessageIdCounter = 1;
    this.documentIdCounter = 1;
    this.voteIdCounter = 1;
    
    // Create a system user for AI responses
    this.createUser({
      username: "system_ai",
      password: "password_not_used_for_system",
      email: "ai@system.com",
      name: "AI Assistant",
      avatar: "",
      department: "System",
      jobTitle: "AI Assistant"
    });
  }

  // User operations
  async getUser(id: number): Promise<User | undefined> {
    return this.users.get(id);
  }

  async getUserByUsername(username: string): Promise<User | undefined> {
    return Array.from(this.users.values()).find(
      (user) => user.username === username,
    );
  }

  async getUserByEmail(email: string): Promise<User | undefined> {
    return Array.from(this.users.values()).find(
      (user) => user.email === email,
    );
  }

  async createUser(insertUser: InsertUser): Promise<User> {
    const id = this.userIdCounter++;
    const createdAt = new Date();
    const user: User = { ...insertUser, id, createdAt };
    this.users.set(id, user);
    return user;
  }

  // Post operations
  async createPost(insertPost: InsertPost): Promise<Post> {
    const id = this.postIdCounter++;
    const createdAt = new Date();
    const post: Post = { 
      ...insertPost, 
      id, 
      createdAt, 
      updatedAt: createdAt,
      views: 0,
      isAnswered: false,
      hasAiAnswer: false
    };
    this.posts.set(id, post);
    return post;
  }

  async getPost(id: number): Promise<Post | undefined> {
    return this.posts.get(id);
  }

  async updatePost(id: number, data: Partial<Post>): Promise<Post | undefined> {
    const post = this.posts.get(id);
    if (!post) return undefined;
    
    const updatedPost = { ...post, ...data, updatedAt: new Date() };
    this.posts.set(id, updatedPost);
    return updatedPost;
  }

  async getPostsByCategory(category: string, page: number, limit: number, sortBy: string): Promise<Post[]> {
    const posts = Array.from(this.posts.values())
      .filter(post => post.category === category);
    
    // Sort posts based on sortBy parameter
    if (sortBy === "recent") {
      posts.sort((a, b) => b.createdAt.getTime() - a.createdAt.getTime());
    } else if (sortBy === "popular") {
      posts.sort((a, b) => b.views - a.views);
    } else if (sortBy === "answered") {
      posts.sort((a, b) => {
        if (a.isAnswered === b.isAnswered) {
          return b.createdAt.getTime() - a.createdAt.getTime();
        }
        return a.isAnswered ? -1 : 1;
      });
    }
    
    // Paginate
    const startIndex = (page - 1) * limit;
    const endIndex = startIndex + limit;
    
    return posts.slice(startIndex, endIndex);
  }

  async countPostsByCategory(category: string): Promise<number> {
    return Array.from(this.posts.values())
      .filter(post => post.category === category).length;
  }

  async searchPosts(query: string): Promise<Post[]> {
    const lowercaseQuery = query.toLowerCase();
    return Array.from(this.posts.values())
      .filter(post => 
        post.title.toLowerCase().includes(lowercaseQuery) || 
        post.content.toLowerCase().includes(lowercaseQuery)
      );
  }

  async searchPostsByCategory(query: string, category: string): Promise<Post[]> {
    const lowercaseQuery = query.toLowerCase();
    return Array.from(this.posts.values())
      .filter(post => 
        post.category === category && 
        (post.title.toLowerCase().includes(lowercaseQuery) || 
         post.content.toLowerCase().includes(lowercaseQuery))
      );
  }

  // Comment operations
  async createComment(insertComment: InsertComment): Promise<Comment> {
    const id = this.commentIdCounter++;
    const createdAt = new Date();
    const comment: Comment = { 
      ...insertComment, 
      id, 
      createdAt,
      upvotes: 0,
      downvotes: 0
    };
    this.comments.set(id, comment);
    return comment;
  }

  async getComment(id: number): Promise<Comment | undefined> {
    return this.comments.get(id);
  }

  async getCommentsByPostId(postId: number): Promise<Comment[]> {
    return Array.from(this.comments.values())
      .filter(comment => comment.postId === postId)
      .sort((a, b) => {
        // Show AI-generated comments first, then by upvotes
        if (a.isAiGenerated !== b.isAiGenerated) {
          return a.isAiGenerated ? -1 : 1;
        }
        
        const aScore = a.upvotes - a.downvotes;
        const bScore = b.upvotes - b.downvotes;
        
        if (aScore !== bScore) {
          return bScore - aScore;
        }
        
        return b.createdAt.getTime() - a.createdAt.getTime();
      });
  }

  async updateCommentVoteCounts(id: number): Promise<void> {
    const comment = this.comments.get(id);
    if (!comment) return;
    
    const votes = Array.from(this.votes.values())
      .filter(vote => vote.commentId === id);
    
    const upvotes = votes.filter(vote => vote.voteType === "upvote").length;
    const downvotes = votes.filter(vote => vote.voteType === "downvote").length;
    
    this.comments.set(id, {
      ...comment,
      upvotes,
      downvotes
    });
  }

  // AI Chat operations
  async createAiChat(insertAiChat: InsertAiChat): Promise<AiChat> {
    const id = this.aiChatIdCounter++;
    const createdAt = new Date();
    const aiChat: AiChat = { ...insertAiChat, id, createdAt };
    this.aiChats.set(id, aiChat);
    return aiChat;
  }

  async getAiChat(id: number): Promise<AiChat | undefined> {
    return this.aiChats.get(id);
  }

  async getAiChatsByUserId(userId: number): Promise<AiChat[]> {
    return Array.from(this.aiChats.values())
      .filter(chat => chat.userId === userId)
      .sort((a, b) => b.createdAt.getTime() - a.createdAt.getTime());
  }

  // AI Chat Message operations
  async createAiChatMessage(insertAiChatMessage: InsertAiChatMessage): Promise<AiChatMessage> {
    const id = this.aiChatMessageIdCounter++;
    const createdAt = new Date();
    const aiChatMessage: AiChatMessage = { ...insertAiChatMessage, id, createdAt };
    this.aiChatMessages.set(id, aiChatMessage);
    return aiChatMessage;
  }

  async getAiChatMessages(chatId: number): Promise<AiChatMessage[]> {
    return Array.from(this.aiChatMessages.values())
      .filter(message => message.chatId === chatId)
      .sort((a, b) => a.createdAt.getTime() - b.createdAt.getTime());
  }

  // Document operations
  async createDocument(insertDocument: InsertDocument): Promise<Document> {
    const id = this.documentIdCounter++;
    const createdAt = new Date();
    const document: Document = { 
      ...insertDocument, 
      id, 
      createdAt,
      status: "processing",
      embeddingId: null
    };
    this.documents.set(id, document);
    return document;
  }

  async getDocument(id: number): Promise<Document | undefined> {
    return this.documents.get(id);
  }

  async updateDocument(id: number, data: Partial<Document>): Promise<Document | undefined> {
    const document = this.documents.get(id);
    if (!document) return undefined;
    
    const updatedDocument = { ...document, ...data };
    this.documents.set(id, updatedDocument);
    return updatedDocument;
  }

  async getDocumentsByCategory(category: string): Promise<Document[]> {
    return Array.from(this.documents.values())
      .filter(document => document.category === category)
      .sort((a, b) => b.createdAt.getTime() - a.createdAt.getTime());
  }

  async getAllDocuments(): Promise<Document[]> {
    return Array.from(this.documents.values())
      .sort((a, b) => b.createdAt.getTime() - a.createdAt.getTime());
  }

  async deleteDocument(id: number): Promise<void> {
    this.documents.delete(id);
  }

  // Vote operations
  async createVote(insertVote: InsertVote): Promise<Vote> {
    const id = this.voteIdCounter++;
    const createdAt = new Date();
    const vote: Vote = { ...insertVote, id, createdAt };
    this.votes.set(id, vote);
    return vote;
  }

  async getVoteByUserAndPost(userId: number, postId: number): Promise<Vote | undefined> {
    return Array.from(this.votes.values())
      .find(vote => vote.userId === userId && vote.postId === postId);
  }

  async getVoteByUserAndComment(userId: number, commentId: number): Promise<Vote | undefined> {
    return Array.from(this.votes.values())
      .find(vote => vote.userId === userId && vote.commentId === commentId);
  }

  async updateVote(id: number, data: Partial<Vote>): Promise<Vote | undefined> {
    const vote = this.votes.get(id);
    if (!vote) return undefined;
    
    const updatedVote = { ...vote, ...data };
    this.votes.set(id, updatedVote);
    return updatedVote;
  }

  async deleteVote(id: number): Promise<void> {
    this.votes.delete(id);
  }
}

export const storage = new MemStorage();

File: python_backend/app/models/storage.py
import bcrypt
import datetime
from typing import Dict, List, Optional, Any, Union

class MemStorage:
    def __init__(self):
        # In-memory storage using dictionaries
        self.users = {}
        self.posts = {}
        self.comments = {}
        self.ai_chats = {}
        self.ai_chat_messages = {}
        self.documents = {}
        self.votes = {}
        
        # ID counters
        self.user_id_counter = 1
        self.post_id_counter = 1
        self.comment_id_counter = 1
        self.ai_chat_id_counter = 1
        self.ai_chat_message_id_counter = 1
        self.document_id_counter = 1
        self.vote_id_counter = 1
        
        # Create system user for AI responses
        self._create_system_user()
    
    def _create_system_user(self):
        """Create a system user for AI-generated content"""
        system_user = {
            "id": 1,
            "username": "ai-assistant",
            "email": "ai@x-ai-forum.com",
            "name": "AI Assistant",
            "password": bcrypt.hashpw("unguessable-password".encode('utf-8'), bcrypt.gensalt()).decode('utf-8'),
            "createdAt": datetime.datetime.now().isoformat()
        }
        self.users[1] = system_user
        self.user_id_counter = 2  # Start regular users at ID 2
    
    # User operations
    def get_user(self, id: int) -> Optional[Dict[str, Any]]:
        """Get user by ID"""
        return self.users.get(id)
    
    def get_user_by_username(self, username: str) -> Optional[Dict[str, Any]]:
        """Get user by username"""
        for user in self.users.values():
            if user["username"] == username:
                return user
        return None
    
    def get_user_by_email(self, email: str) -> Optional[Dict[str, Any]]:
        """Get user by email"""
        for user in self.users.values():
            if user["email"] == email:
                return user
        return None
    
    def create_user(self, user_data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new user"""
        user_id = self.user_id_counter
        self.user_id_counter += 1
        
        created_at = datetime.datetime.now().isoformat()
        
        user = {**user_data, "id": user_id, "createdAt": created_at}
        self.users[user_id] = user
        
        return user
    
    # Post operations
    def create_post(self, post_data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new forum post"""
        post_id = self.post_id_counter
        self.post_id_counter += 1
        
        created_at = datetime.datetime.now().isoformat()
        
        post = {
            **post_data,
            "id": post_id,
            "createdAt": created_at,
            "views": 0,
            "isAnswered": False,
            "hasAiAnswer": False
        }
        
        self.posts[post_id] = post
        
        # Add user info if needed
        user = self.get_user(post_data["userId"])
        if user:
            post["user"] = {
                "id": user["id"],
                "username": user["username"],
                "name": user["name"],
                "avatar": user.get("avatar"),
                "jobTitle": user.get("jobTitle")
            }
        
        return post
    
    def get_post(self, id: int) -> Optional[Dict[str, Any]]:
        """Get post by ID"""
        post = self.posts.get(id)
        if not post:
            return None
        
        # Add user info if needed
        if "user" not in post and "userId" in post:
            user = self.get_user(post["userId"])
            if user:
                post["user"] = {
                    "id": user["id"],
                    "username": user["username"],
                    "name": user["name"],
                    "avatar": user.get("avatar"),
                    "jobTitle": user.get("jobTitle")
                }
        
        # Add comment count
        post["_count"] = {
            "comments": len([c for c in self.comments.values() if c.get("postId") == id])
        }
        
        return post
    
    def update_post(self, id: int, data: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Update a post"""
        post = self.posts.get(id)
        if not post:
            return None
        
        for key, value in data.items():
            post[key] = value
        
        post["updatedAt"] = datetime.datetime.now().isoformat()
        return post
    
    def get_posts_by_category(self, category: str, page: int, limit: int, sort_by: str) -> List[Dict[str, Any]]:
        """Get posts by category with pagination and sorting"""
        # Filter by category
        filtered_posts = [p for p in self.posts.values() if p.get("category") == category]
        
        # Apply sorting
        if sort_by == "recent":
            filtered_posts.sort(key=lambda p: p.get("createdAt", ""), reverse=True)
        elif sort_by == "popular":
            filtered_posts.sort(key=lambda p: p.get("views", 0), reverse=True)
        elif sort_by == "unanswered":
            filtered_posts = [p for p in filtered_posts if not p.get("isAnswered", False)]
            filtered_posts.sort(key=lambda p: p.get("createdAt", ""), reverse=True)
        
        # Calculate pagination
        start = (page - 1) * limit
        end = start + limit
        paginated_posts = filtered_posts[start:end]
        
        # Add user info and comment counts to each post
        for post in paginated_posts:
            if "user" not in post and "userId" in post:
                user = self.get_user(post["userId"])
                if user:
                    post["user"] = {
                        "id": user["id"],
                        "username": user["username"],
                        "name": user["name"],
                        "avatar": user.get("avatar"),
                        "jobTitle": user.get("jobTitle")
                    }
            
            # Add comment count
            post["_count"] = {
                "comments": len([c for c in self.comments.values() if c.get("postId") == post["id"]])
            }
        
        return paginated_posts
    
    def count_posts_by_category(self, category: str) -> int:
        """Count posts in a category"""
        return len([p for p in self.posts.values() if p.get("category") == category])
    
    def search_posts(self, query: str) -> List[Dict[str, Any]]:
        """Search posts by title and content"""
        query_lower = query.lower()
        matched_posts = []
        
        for post in self.posts.values():
            title = post.get("title", "").lower()
            content = post.get("content", "").lower()
            
            if query_lower in title or query_lower in content:
                matched_posts.append(post)
        
        return matched_posts
    
    def search_posts_by_category(self, query: str, category: str) -> List[Dict[str, Any]]:
        """Search posts by title and content within a category"""
        query_lower = query.lower()
        matched_posts = []
        
        for post in self.posts.values():
            if post.get("category") != category:
                continue
                
            title = post.get("title", "").lower()
            content = post.get("content", "").lower()
            
            if query_lower in title or query_lower in content:
                matched_posts.append(post)
        
        return matched_posts
    
    # Comment operations
    def create_comment(self, comment_data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new comment"""
        comment_id = self.comment_id_counter
        self.comment_id_counter += 1
        
        created_at = datetime.datetime.now().isoformat()
        
        comment = {
            **comment_data,
            "id": comment_id,
            "createdAt": created_at,
            "upvotes": 0,
            "downvotes": 0
        }
        
        self.comments[comment_id] = comment
        
        # Add user info
        user = self.get_user(comment_data["userId"])
        if user:
            comment["user"] = {
                "id": user["id"],
                "username": user["username"],
                "name": user["name"],
                "avatar": user.get("avatar"),
                "jobTitle": user.get("jobTitle")
            }
        
        return comment
    
    def get_comment(self, id: int) -> Optional[Dict[str, Any]]:
        """Get comment by ID"""
        comment = self.comments.get(id)
        if not comment:
            return None
            
        # Add user info if needed
        if "user" not in comment and "userId" in comment:
            user = self.get_user(comment["userId"])
            if user:
                comment["user"] = {
                    "id": user["id"],
                    "username": user["username"],
                    "name": user["name"],
                    "avatar": user.get("avatar"),
                    "jobTitle": user.get("jobTitle")
                }
        
        return comment
    
    def get_comments_by_post_id(self, post_id: int) -> List[Dict[str, Any]]:
        """Get all comments for a post"""
        filtered_comments = [c for c in self.comments.values() if c.get("postId") == post_id]
        
        # Sort by creation date
        filtered_comments.sort(key=lambda c: c.get("createdAt", ""))
        
        # Add user info to each comment
        for comment in filtered_comments:
            if "user" not in comment and "userId" in comment:
                user = self.get_user(comment["userId"])
                if user:
                    comment["user"] = {
                        "id": user["id"],
                        "username": user["username"],
                        "name": user["name"],
                        "avatar": user.get("avatar"),
                        "jobTitle": user.get("jobTitle")
                    }
        
        return filtered_comments
    
    def update_comment_vote_counts(self, id: int) -> None:
        """Update upvote and downvote counts for a comment"""
        comment = self.comments.get(id)
        if not comment:
            return
        
        # Count votes for this comment
        upvotes = 0
        downvotes = 0
        
        for vote in self.votes.values():
            if vote.get("commentId") == id:
                if vote.get("voteType") == "upvote":
                    upvotes += 1
                elif vote.get("voteType") == "downvote":
                    downvotes += 1
        
        # Update comment
        comment["upvotes"] = upvotes
        comment["downvotes"] = downvotes
    
    # AI Chat operations
    def create_ai_chat(self, chat_data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new AI chat"""
        chat_id = self.ai_chat_id_counter
        self.ai_chat_id_counter += 1
        
        created_at = datetime.datetime.now().isoformat()
        
        chat = {**chat_data, "id": chat_id, "createdAt": created_at}
        self.ai_chats[chat_id] = chat
        
        return chat
    
    def get_ai_chat(self, id: int) -> Optional[Dict[str, Any]]:
        """Get AI chat by ID"""
        return self.ai_chats.get(id)
    
    def get_ai_chats_by_user_id(self, user_id: int) -> List[Dict[str, Any]]:
        """Get all AI chats for a user"""
        filtered_chats = [c for c in self.ai_chats.values() if c.get("userId") == user_id]
        filtered_chats.sort(key=lambda c: c.get("createdAt", ""), reverse=True)
        
        # Add last message preview
        for chat in filtered_chats:
            # Get last message
            chat_messages = [m for m in self.ai_chat_messages.values() if m.get("chatId") == chat["id"]]
            if chat_messages:
                chat_messages.sort(key=lambda m: m.get("createdAt", ""), reverse=True)
                last_message = chat_messages[0]
                
                # Truncate content
                content = last_message.get("content", "")
                if len(content) > 50:
                    content = content[:50] + "..."
                
                chat["lastMessage"] = {
                    "content": content,
                    "createdAt": last_message.get("createdAt"),
                    "isUserMessage": last_message.get("isUserMessage", False)
                }
        
        return filtered_chats
    
    # AI Chat Message operations
    def create_ai_chat_message(self, message_data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new AI chat message"""
        message_id = self.ai_chat_message_id_counter
        self.ai_chat_message_id_counter += 1
        
        created_at = datetime.datetime.now().isoformat()
        
        message = {**message_data, "id": message_id, "createdAt": created_at}
        self.ai_chat_messages[message_id] = message
        
        return message
    
    def get_ai_chat_messages(self, chat_id: int) -> List[Dict[str, Any]]:
        """Get all messages for an AI chat"""
        filtered_messages = [m for m in self.ai_chat_messages.values() if m.get("chatId") == chat_id]
        filtered_messages.sort(key=lambda m: m.get("createdAt", ""))
        return filtered_messages
    
    # Document operations
    def create_document(self, document_data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new document"""
        document_id = self.document_id_counter
        self.document_id_counter += 1
        
        created_at = datetime.datetime.now().isoformat()
        
        document = {
            **document_data,
            "id": document_id,
            "createdAt": created_at,
            "status": "pending"
        }
        
        self.documents[document_id] = document
        
        # Add uploaded by user info
        user = self.get_user(document_data["uploadedById"])
        if user:
            document["uploadedBy"] = {
                "id": user["id"],
                "username": user["username"],
                "name": user["name"]
            }
        
        return document
    
    def get_document(self, id: int) -> Optional[Dict[str, Any]]:
        """Get document by ID"""
        document = self.documents.get(id)
        if not document:
            return None
            
        # Add user info if needed
        if "uploadedBy" not in document and "uploadedById" in document:
            user = self.get_user(document["uploadedById"])
            if user:
                document["uploadedBy"] = {
                    "id": user["id"],
                    "username": user["username"],
                    "name": user["name"]
                }
        
        return document
    
    def update_document(self, id: int, data: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Update a document"""
        document = self.documents.get(id)
        if not document:
            return None
        
        for key, value in data.items():
            document[key] = value
        
        document["updatedAt"] = datetime.datetime.now().isoformat()
        return document
    
    def get_documents_by_category(self, category: str) -> List[Dict[str, Any]]:
        """Get documents by category"""
        filtered_documents = [d for d in self.documents.values() if d.get("category") == category]
        filtered_documents.sort(key=lambda d: d.get("createdAt", ""), reverse=True)
        
        # Add user info to each document
        for document in filtered_documents:
            if "uploadedBy" not in document and "uploadedById" in document:
                user = self.get_user(document["uploadedById"])
                if user:
                    document["uploadedBy"] = {
                        "id": user["id"],
                        "username": user["username"],
                        "name": user["name"]
                    }
        
        return filtered_documents
    
    def get_all_documents(self) -> List[Dict[str, Any]]:
        """Get all documents"""
        documents = list(self.documents.values())
        documents.sort(key=lambda d: d.get("createdAt", ""), reverse=True)
        
        # Add user info to each document
        for document in documents:
            if "uploadedBy" not in document and "uploadedById" in document:
                user = self.get_user(document["uploadedById"])
                if user:
                    document["uploadedBy"] = {
                        "id": user["id"],
                        "username": user["username"],
                        "name": user["name"]
                    }
        
        return documents
    
    def delete_document(self, id: int) -> None:
        """Delete a document"""
        if id in self.documents:
            del self.documents[id]
    
    # Vote operations
    def create_vote(self, vote_data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new vote"""
        vote_id = self.vote_id_counter
        self.vote_id_counter += 1
        
        created_at = datetime.datetime.now().isoformat()
        
        vote = {**vote_data, "id": vote_id, "createdAt": created_at}
        self.votes[vote_id] = vote
        
        return vote
    
    def get_vote_by_user_and_post(self, user_id: int, post_id: int) -> Optional[Dict[str, Any]]:
        """Get a vote by user and post"""
        for vote in self.votes.values():
            if vote.get("userId") == user_id and vote.get("postId") == post_id:
                return vote
        return None
    
    def get_vote_by_user_and_comment(self, user_id: int, comment_id: int) -> Optional[Dict[str, Any]]:
        """Get a vote by user and comment"""
        for vote in self.votes.values():
            if vote.get("userId") == user_id and vote.get("commentId") == comment_id:
                return vote
        return None
    
    def update_vote(self, id: int, data: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Update a vote"""
        vote = self.votes.get(id)
        if not vote:
            return None
        
        for key, value in data.items():
            vote[key] = value
        
        vote["updatedAt"] = datetime.datetime.now().isoformat()
        return vote
    
    def delete_vote(self, id: int) -> None:
        """Delete a vote"""
        if id in self.votes:
            del self.votes[id]

# Create a single instance of the storage
storage = MemStorage()